File: DSA-3/Arrays/stock-buy-and-sell.cpp
/***

https://www.geeksforgeeks.org/stock-buy-sell/

****/

/*This is a function problem.You only need to complete the function given below*/
// solution structure
/* struct Interval
 {
     int buy;
     int sell;
 };*/
// This function finds the buy sell schedule for maximum profit
void stockBuySell(int price[], int n)
{
    if(n == 1){
        return;
    }
    // Your code here
    int i = 0;
    int count = 0;
    
    Interval res[n/2 + 1];
    
    while(i < n-1){
        
        while((i < n -1 ) && price[i + 1] <= price[i]){
            i++;
        }
        
        if(i == n-1){
            break;
        }
        
        res[count].buy = i++;
        
        while((i < n) && price[i] >= price[i-1]){
            i++;
        }
        
        res[count].sell = i - 1;
        count++;
    }
    
    if(count == 0){
        cout<<"No Profit";
    }else{
        
        for(int i=0; i<count; i++){
            cout<<"("<<res[i].buy<<" "<<res[i].sell<<")"<<" ";
        }
    }
    
    return;
}

File: Company Wise Practice/Goldman Sachs/stock-buy-and-sell.java
/*package whatever //do not write package name here */
/********

https://www.geeksforgeeks.org/stock-buy-sell/

********/
import java.util.*;
import java.lang.*;
import java.io.*;

class Interval{
    int buy;
    int sell;
}

class GFG {
    
    public static void stockBuySell(int arr[],int n){
        
        int count = 0;
        
        if(n == 1){
            System.out.println("No Profit");
        }else{
            
            List<Interval> list = new ArrayList<>();
            
            int i = 0;
            while(i < n - 1){
                
                while((i < n - 1) && (arr[i+1] <= arr[i])){
                    i++;
                }
                
                if(i == n-1){
                    break;
                }
                
                Interval temp = new Interval();
                temp.buy = i++;
                
                while( (i < n) && (arr[i] >= arr[i-1])){
                    i++;
                }
                
                temp.sell = i - 1;
                list.add(temp);
                count++;
            }
            
            if(count == 0){
                System.out.println("No Profit");
            }else{
               
               for(int j=0; j<list.size(); j++){
                   System.out.print("("+list.get(j).buy+" "+list.get(j).sell+") ");
               }
                System.out.println();
            }
        }
    }
    
	public static void main (String[] args) {
		//code
		Scanner sc =new Scanner(System.in);
		Integer t = sc.nextInt();
		
		while(t-- > 0){
		    
		    Integer n = sc.nextInt();
		    
		    int arr[] = new int[n];
		    
		    for(int i=0; i<n; i++){
		        arr[i] = sc.nextInt();
		    }
		    stockBuySell(arr,n);
		    
		}
		
		sc.close();
	}
}